src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -124,30 +135,43 @@ definition index :: "('i \<Rightarrow> 'q) \<Rightarrow> 'q \<Rightarrow> 'i lis
 definition first_intersection_point :: "('i \<Rightarrow> 'q) \<Rightarrow> 'q list \<Rightarrow> 'i list \<rightharpoonup> 'q" where
   [simp]: "first_intersection_point \<psi> qs C  = find (\<lambda>q. q \<in> \<psi> ` set C) qs"
 
+definition last_intersection_point :: "('i \<Rightarrow> 'q) \<Rightarrow> 'i list \<Rightarrow> 'i list \<Rightarrow> 'q" where
+  [simp]: "last_intersection_point \<psi> C C' = (ARG_MAX (\<lambda>q. index \<psi> q C) q. q \<in> \<psi> ` (set C \<inter> set C'))"
+
 corec server_program where
-  "server_program \<psi> C k b =
+  "server_program \<psi> C k b a =
     \<down> M. (partial_case M of
       Done \<Rightarrow>
         \<bottom> |
       Cont (FindIntersect qs) \<Rightarrow>
         (case first_intersection_point \<psi> qs C of
           None \<Rightarrow>
             \<up> Cont IntersectNotFound;
-            server_program \<psi> C k b |
+            server_program \<psi> C k b a |
           Some q \<Rightarrow>
             \<up> Cont (IntersectFound q);
-            server_program \<psi> C (index \<psi> q C) True
+            server_program \<psi> C (index \<psi> q C) True a
         ) |
       Cont RequestNext \<Rightarrow>
         if b then
           \<up> Cont (RollBackward (\<psi> (C ! k)));
-          server_program \<psi> C k False
+          server_program \<psi> C k False a
         else if Suc k < length C then
           \<up> Cont (RollForward (C ! Suc k));
-          server_program \<psi> C (Suc k) b
-        else \<comment> \<open>only for this initial implementation\<close>
+          server_program \<psi> C (Suc k) b a
+        else
           \<up> Cont AwaitReply;
-          server_program \<psi> C k b
+          a \<rightarrow> C'. (
```

Comment:
How about the following layout?

```
u → C'.
(
let q = ... in
if q = ... then
  ...
else
  ...
)
```

Reply:
I think that the stuff between the parentheses should be indented relative to the parentheses, as it is subordinate to the whole parenthesized expression. Otherwise, it looks good to me.

Reply:
Fixed in 73ef50b358e084c25fa595147ef5881a6e2b21cb, please verify.